scripts/timelapse-colorizer-data/generate_data.py

```diff
@@ -139,107 +117,62 @@ def make_frames(grouped_frames, output_dir, dataset, scale: float):
+    tracks = a[TRACK_ID_COLUMN].to_numpy()
+    times = a[TIMES_COLUMN].to_numpy()
     centroids_x = a["centroid_x"].to_numpy()
     centroids_y = a["centroid_y"].to_numpy()
-    centroids_stacked = np.ravel(np.dstack([centroids_x, centroids_y]))
-    centroids_stacked = centroids_stacked * scale
-    centroids_stacked = centroids_stacked.astype(int)
 
-    centroids_json = {"data": centroids_stacked.tolist()}
-    with open(outpath + "/centroids.json", "w") as f:
-        json.dump(centroids_json, f)
-
-    logging.info("Writing feature json...")
-    for i in range(nfeatures):
-        f = a[features[i]].to_numpy()
-        fmin = np.nanmin(f)
-        fmax = np.nanmax(f)
+    feature_data = []
+    for i in range(len(features)):
         # TODO normalize output range excluding outliers?
-        js = {"data": f.tolist(), "min": fmin, "max": fmax}
-        with open(outpath + "/feature_" + str(i) + ".json", "w") as f:
-            json.dump(js, f)
-    logging.info("Done writing features.")
+        f = a[features[i]].to_numpy()
+        feature_data.append(f)
+
+    writer.write_feature_data(
+        feature_data,
+        tracks,
+        times,
+        centroids_x,
+        centroids_y,
+        outliers,
+    )
 
 
 def make_dataset(output_dir="./data/", dataset="baby_bear", do_frames=True, scale=1):
-    os.makedirs(os.path.join(output_dir, dataset), exist_ok=True)
+    """Make a new dataset from the given data, and write the complete dataset
+    files to the given output directory.
+    """
+    writer = ColorizerDatasetWriter(output_dir, dataset, scale=scale)
 
     # use nucmorph to load data
     datadir, figdir = create_base_directories(dataset)
     pixsize = get_dataset_pixel_size(dataset)
 
-    # a is the full dataset!
-    a = load_dataset(dataset, datadir=None)
+    full_dataset = load_dataset(dataset, datadir=None)
     logging.info("Loaded dataset '" + str(dataset) + "'.")
 
-    columns = ["track_id", "index_sequence", "seg_full_zstack_path", "label_img"]
-    # b is the reduced dataset
-    b = a[columns]
-    b = b.reset_index(drop=True)
-    b["initialIndex"] = b.index.values
-
-    grouped_frames = b.groupby("index_sequence")
-    # get a single path from each time in the set.
-    # frames = grouped_frames.apply(lambda df: df.sample(1))
+    # Make a reduced dataframe grouped by time (frame number).
+    columns = [TRACK_ID_COLUMN, TIMES_COLUMN, SEGMENTED_IMAGE_COLUMN, OBJECT_ID_COLUMN]
+    reduced_dataset = full_dataset[columns]
+    reduced_dataset = reduced_dataset.reset_index(drop=True)
+    reduced_dataset[INITIAL_INDEX_COLUMN] = reduced_dataset.index.values
+    grouped_frames = reduced_dataset.groupby(TIMES_COLUMN)
 
+    # Make the features, frame data, and manifest.
     nframes = len(grouped_frames)
-
     features = ["NUC_shape_volume_lcc", "NUC_position_depth"]
-    make_features(a, features, output_dir, dataset, scale)
-
+    make_features(full_dataset, features, writer)
```

Comment:
nit: i still wonder if this would be more lightweight if we passed reduced_dataset in (making sure it has the feature columns already) 🤷

Reply:
I think the performance difference might be pretty negligible? This part of the code runs pretty fast (1-2sec) compared to `make_frames`, which takes several hours.